src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/MessageBody.cs

```diff
@@ -378,6 +383,14 @@ protected override bool Read(ReadableBuffer readableBuffer, WritableBuffer writa
 
                 return _inputLength == 0;
             }
+
+            protected override void OnReadStart()
+            {
+                if (_contentLength > _context.MaxRequestBodySize)
+                {
+                    _context.RejectRequest(RequestRejectionReason.RequestBodyTooLarge);
```

Comment:
I was tempted to change the name of RejectRequest to ThrowBadHttpRequestException as part of this change. I could still do that if we're in agreement that it makes this and all the other calling code at least slightly more readable.